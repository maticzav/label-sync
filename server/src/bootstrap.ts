import { compile } from 'handlebars'

import { mapEntries } from './data/dict'
import { GHTree } from './github'

/**
 * Populates the template with repositories.
 * @param tree
 */
export function populateTempalte(
  tree: GHTree,
  data: { repository: string; repositories: { name: string }[] },
): GHTree {
  return mapEntries(tree, (file) => compile(file)(data))
}
