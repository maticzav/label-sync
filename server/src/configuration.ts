import * as t from 'io-ts'
import { PathReporter } from 'io-ts/lib/PathReporter'
import yaml from 'js-yaml'
import * as os from 'os'

import { mapEntries } from './data/dict'

/**
 * Configuration repository is the repository which LabelSync
 * uses to determine the configuration of the service.
 *
 * @param organization
 */
export const getLSConfigRepoName = (owner: string) => `${owner}-labelsync`

/**
 * Configuration file path determines the path of the file in the repositoy
 * that we use to gather configuration information from.
 *
 * It should always be in YAML format.
 */
export const LS_CONFIG_PATH = 'labelsync.yml'

/**
 * Label represents the central unit of LabelSync. Each label
 * can have multiple siblings that are meaningfully related to
 * a label itself, multiple hooks that trigger different actions.
 */
const LSCLabel = t.intersection([
  t.type({
    color: t.string,
  }),
  t.partial({
    description: t.string,
    siblings: t.array(t.string),
    alias: t.array(t.string),
  }),
])
export type LSCLabel = t.TypeOf<typeof LSCLabel>

const LSCLabelName = t.string
export type LSCLabelName = t.TypeOf<typeof LSCLabelName>

/**
 * Repository configuration for how LabelSync should sync it.
 */
const LSCRepositoryConfiguration = t.partial({
  removeUnconfiguredLabels: t.boolean,
})
export interface LSCRepositoryConfiguration
  extends t.TypeOf<typeof LSCRepositoryConfiguration> {}

/**
 * Repository represents a single Github repository.
 * When configured as `strict` it will delete any surplus of labels
 * in the repository.
 */
const LSCRepository = t.intersection([
  t.partial({
    config: LSCRepositoryConfiguration,
  }),
  t.type({
    labels: t.record(LSCLabelName, LSCLabel),
  }),
])
export interface LSCRepository extends t.TypeOf<typeof LSCRepository> {}
const LSCRepositoryName = t.string
export type LSCRepositoryName = t.TypeOf<typeof LSCRepositoryName>

/**
 * Configuration represents an entire configuration for all
 * LabelSync tools that an organisation is using.
 */
const LSCConfiguration = t.type({
  repos: t.record(LSCRepositoryName, LSCRepository),
})
export interface LSCConfiguration extends t.TypeOf<typeof LSCConfiguration> {}

/**
 * Parses a string into a configuration object.
 * @param input
 */
export function parseConfig(
  input: string,
  logger?: (err: any) => any,
): [string] | [null, LSCConfiguration] {
  try {
    const object = yaml.safeLoad(input)
    const config = LSCConfiguration.decode(object)

    if (config._tag === 'Left') {
      return [PathReporter.report(config).join(os.EOL)]
    }
    return [null, fixConfig(config.right)]
  } catch (err) /* istanbul ignore next */ {
    if (logger) logger(err)
    return [err.message]
  }
}

/**
 * Catches configuration issues that io-ts couldn't catch.
 *
 * @param config
 */
function fixConfig(config: LSCConfiguration): LSCConfiguration {
  const repos = mapEntries(config.repos, fixRepoConfig)

  return {
    ...config,
    repos: repos,
  }
}

/**
 * Fixes issues that io-ts couldn't catch.
 *
 * @param config
 */
function fixRepoConfig(config: LSCRepository): LSCRepository {
  const labels = mapEntries<LSCLabel, LSCLabel>(config.labels, label => ({
    ...label,
    color: fixLabelColor(label.color),
  }))

  return {
    ...config,
    labels,
  }
}

/**
 * Removes hash from the color.
 *
 * @param color
 */
function fixLabelColor(color: string): string {
  if (color.startsWith('#')) return color.slice(1)
  return color
}
