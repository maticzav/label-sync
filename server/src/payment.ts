import Stripe from 'stripe'
// import { PrismaClient } from '@prisma/client'

export class Payments {
  client: Stripe
  endpointSecret: string

  constructor(apiKey: string, endpointSecret: string) {
    this.client = new Stripe(apiKey, {
      apiVersion: '2020-03-02',
    })
    this.endpointSecret = endpointSecret
    // this.db = prisma
  }

  // /**
  //  * Saves a purchase.
  //  */
  // async purchase(owner: string) {
  //   return this.db.purchase.create({})
  // }

  /**
   * Consturcts a purchase event.
   */
  async constructEvent(payload: string | Buffer, header: string | Buffer) {
    return this.client.webhooks.constructEvent(
      payload,
      header,
      this.endpointSecret,
    )
  }

  /**
   * Constructs a new checkout session.
   */
  async getSession(owner: string) {
    return this.client.checkout.sessions.create({
      payment_method_types: ['card'],
      subscription_data: {
        items: [
          {
            plan: 'plan_H2NVZcTwOuqJkT',
          },
        ],
      },
      metadata: {
        owner: owner,
      },
      success_url: 'https://github.com/apps/labelsync-manager',
      cancel_url: 'https://label-sync.com',
    })
  }
}
