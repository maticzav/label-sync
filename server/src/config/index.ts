import { Plan } from '@prisma/client'

import * as parser from './parser'
import { checkConfiguration } from './checks/configuration'
import { checkPlan } from './checks/plan'
import { getLSConfigRepoName } from './constants'

import * as t from './types'

/**
 * This file bundles up all functions related to configuration.
 */

export * from './constants'
export * from './types'

/**
 * Parses the configuration.
 */
export function parse(params: {
  plan: Plan
  input: string
}): [string] | [null, t.LSCConfiguration] {
  const result = parser.parse(params.input, [
    checkPlan(params.plan),
    checkConfiguration,
  ])

  if (result.success) {
    return [null, result.config]
  }
  return [result.error]
}

// MARK: - Utility functions

/**
 * Returns a list of repositories in configuration without wildcard config.
 */
export function configRepos(config: t.LSCConfiguration): string[] {
  let repos: string[] = []

  for (let repo of config.repos.keys()) {
    repo = repo.trim()
    if (repo !== '*') repos.push(repo)
  }
  return repos
}

/**
 * Tells whether the repository is the configuration
 * repository for the account.
 */
export function isConfigRepo(account: string, repo: string): boolean {
  return getLSConfigRepoName(account) === repo.toLowerCase()
}
