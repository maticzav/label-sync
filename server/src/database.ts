/**
 * This file contains the datasource that connects to the database.
 */

import { PrismaClient } from '@prisma/client'
import * as prisma from '@prisma/client'

import * as t from 'luxon'

import * as c from './config'
import { Maybe } from './data/maybe'

export class Data {
  // MARK: - State

  private prisma: PrismaClient

  // MARK: - Constructor

  constructor(prisma: PrismaClient) {
    this.prisma = prisma
  }

  // MARK: - Methods

  /**
   * Returns the configuration of a particular installation.
   */
  async config(owner: string, repo: string): Promise<Maybe<c.LSCRepository>> {
    const [config] = await this.prisma.repositoryConfiguration.findMany({
      where: {
        configuration: { active: { owner: owner } },
        repository: { in: [repo, '*'] },
      },
      orderBy: {
        repository: 'asc',
      },
      include: {
        labels: true,
      },
    })

    if (config === undefined) return null

    let configuration: c.LSCRepository = {
      config: {
        removeUnconfiguredLabels: config.removeUnconfiguredLabels,
      },
      labels: new Map(),
    }

    for (const label of config.labels) {
      configuration.labels.set(label.name, {
        color: label.color,
        description: label.description,
        alias: label.alias,
        siblings: label.siblings,
      })
    }

    return configuration
  }

  /**
   * Returns the user plan.
   */
  async plan(owner: string): Promise<prisma.Plan> {
    const now = t.DateTime.now()

    const installation = await this.prisma.installation.findUnique({
      where: { owner: owner },
    })

    /**
     * Check that we have an installation and that plan is still valid.
     */
    if (installation === null) return 'FREE'
    if (t.DateTime.fromJSDate(installation.periodEndsAt) < now) return 'FREE'

    return installation.plan
  }

  async install(): Promise<void> {}
}
