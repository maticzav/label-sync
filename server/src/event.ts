/**
 * This file contains a specification for a general event that
 * should be used for adding event handlers to the app.
 *
 * This might seem like a lot of boilerplate; the idea behind separation
 * is that we get a clear instance that works in deterministic way based on
 * the input. You can think of an event as code executing in a very limited
 * environemnt. This file outlines the functions available in it.
 */

import { Probot } from 'probot'
import { ProbotWebhooks } from 'probot/lib/types'

import { Data } from './database'

// MARK: - Spec

// export type Event<E extends WebhookEvents> = (
//   event: E | E[],
//   callback: HandlerFunction<E, U>,
// ) => void

export type Handler = (on: ProbotWebhooks['on'], sources: Sources) => void

/**
 * Sources outline third-party functions available to the event code
 * in the execution environment.
 */
export type Sources = {
  data: IData
}

export interface IData {
  /**
   * Returns the installation associated with a given owner.
   */
  plan: Data['plan']

  /**
   * Returns information about a configuration for a given repository.
   */
  config: Data['config']

  /**
   * Updates the last installation for a given owner.
   */
  install: Data['install']
}

// MARK: - Methods

type BindParams = {
  handlers: Handler[]
  probot: Probot
  sources: Sources
}

/**
 * Binds handlers to sources.
 */
export function bind(params: BindParams) {}
