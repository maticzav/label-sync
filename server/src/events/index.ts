import { Handler } from '../event'

import { handler as installation } from './installation'
import { handler as label } from './label'
import { handler as pullrequest } from './pullrequest'
import { handler as push } from './push'
import { handler as repository } from './repository'

/**
 * Aggregates all the events.
 */
export { handler as installation } from './installation'
export { handler as label } from './label'
export { handler as pullrequest } from './pullrequest'
export { handler as push } from './push'
export { handler as repository } from './repository'

export const handlers: Handler[] = [
  installation,
  label,
  pullrequest,
  push,
  repository,
]
