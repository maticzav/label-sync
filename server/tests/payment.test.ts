import { Period } from '@prisma/client'
import moment from 'moment'

import { Payments } from '../src/payment'

describe('payments:', () => {
  let payments: Payments

  beforeAll(() => {
    payments = new Payments('key', 'secret')
  })

  test('sets right expiration date:', () => {
    expect(
      366 -
        moment(payments.getExpirationDateFromNow('ANNUALLY')).diff(
          moment(),
          'day',
        ),
    ).toBeLessThan(3)
    expect(
      31 -
        moment(payments.getExpirationDateFromNow('MONTHLY')).diff(
          moment(),
          'day',
        ),
    ).toBeLessThan(3)
  })

  test('active plans:', () => {
    const plans: Period[] = ['MONTHLY', 'ANNUALLY']
    expect(plans.map(payments.getPlanForPeriod)).toMatchSnapshot()
  })
})
