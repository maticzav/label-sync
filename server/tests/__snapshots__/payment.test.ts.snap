// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`payments: active plans: 1`] = `
Array [
  Object {
    "plan": "plan_HCkojOBbK8hFh6",
  },
  Object {
    "plan": "plan_HCkpZId8BCi7cI",
  },
]
`;
